calculator/calculator.py

```diff
@@ -1,15 +1,115 @@
 import re
+import math
+from typing import Union
+from statistics import mean
 
 
-class Calculator:
+_NUMBER_REGEX = r'\d+(?:\.\d+)?'
+_NUMBER_NEGATIVE_OPT_REGEX = r'-?\d+(?:\.\d+)?'
 
-    @classmethod
-    def calculate(cls, expression=str):
-        if type(expression) is not str:
-            raise TypeError("Expression input should be of str type.")
+OPERATION_REGEX_INDEX = 0
+OPERATION_HANDLER_INDEX = 1
 
-        # TODO
 
+def calculate(expression: str) -> Union[int, float]:
+    if type(expression) is not str:
+        raise TypeError("Expression input should be of str type.")
 
+    return _calculate(expression.replace(" ", ""))
 
+def _calculate(expression: str) -> Union[int, float]:
+    if _is_end_of_calculation(expression):
+        return expression
 
+    for _, value in _TOKEN_SPECIFICATION.items():
+        tmp_result = _handle(expression, value)
+        if tmp_result is not None:
+            return _calculate(tmp_result)
+
+    raise ValueError(f'Expression: {expression} is not legal!')
+
+def _one_variable_operation(operation: str, is_negative_legal: bool = False) -> str:
+    if is_negative_legal is False:
+        return f'(?P<NUMBER>{_NUMBER_REGEX}){operation}'
+    else:
+        return f'(?P<NUMBER>{_NUMBER_NEGATIVE_OPT_REGEX}){operation}'
+
+def _two_variable_operation(operation: str, is_negative_legal: bool = False) -> str:
+    if is_negative_legal is False:
+        return f'(?P<FIRST_NUMBER>{_NUMBER_REGEX}){operation}' \
+               f'(?P<SECOND_NUMBER>{_NUMBER_REGEX})'
+    else:
+        return f'(?P<FIRST_NUMBER>{_NUMBER_NEGATIVE_OPT_REGEX}){operation}' \
+               f'(?P<SECOND_NUMBER>{_NUMBER_NEGATIVE_OPT_REGEX})'
+
+def _is_end_of_calculation(expression: str) -> bool:
+    return re.fullmatch(f'{_NUMBER_NEGATIVE_OPT_REGEX}', expression) is not None
+
+def _handle(expression: str, operation_details) -> str:
+    token = re.search(operation_details[OPERATION_REGEX_INDEX], expression)
+    if token:
+        return re.sub(token.re, operation_details[OPERATION_HANDLER_INDEX](token), expression, 1)
+
+def _handle_parenthesis(token) -> str:
+    return str(_calculate(token.group("EXTRACTED_EXPRESSION")))
+
+def _handle_factorial(token) -> str:
+    return str(math.factorial(float(token.group("NUMBER"))))
+
+def _handle_negate(token) -> str:
+    return str(-1 * float(token.group("NUMBER")))
```

Comment:
It seems that unary operation are handled the same except the operation.
(You extract NUMBER, cast to float, and perform operation X).
Make it generic!